crates/matrix-sdk/src/oidc.rs

```diff
@@ -0,0 +1,966 @@
+    ) -> Result<AccessTokenResponse, OidcError> {
+        let response =
+            self.oidc.finish_authorization_with_authorization_code(auth_response).await?;
+        self.oidc.load_session_meta().await?;
+
+        Ok(response)
+    }
+
+    /// Login via the Client Credentials flow.
+    ///
+    /// # Arguments
+    ///
+    /// * `scope` - The scope to authorize.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// use matrix_sdk::{Client, Error};
+    /// # use futures::executor::block_on;
+    /// # use url::Url;
+    /// # let homeserver = Url::parse("https://example.com").unwrap();
+    /// # block_on(async {
+    /// # let client = Client::new(homeserver).await?;
+    /// let oidc = client.oidc();
+    ///
+    /// let _response = oidc.login_with_client_credentials().await?;
+    ///
+    /// // The access and refresh tokens can be persisted either from the response,
+    /// // or from one of the `Client::session_tokens*` methods.
+    ///
+    /// // You can now make any request compatible with the requested scope.
+    /// let _me = client.whoami().await?;
+    /// # anyhow::Ok(()) })
+    /// ```
+    #[instrument(target = "matrix_sdk::client", skip_all)]
+    pub async fn login_with_client_credentials(self) -> Result<AccessTokenResponse, OidcError> {
+        let issuer = self.oidc.issuer().await.ok_or(OidcError::MissingAuthenticationIssuer)?;
+        info!(issuer, "Logging in via OpenID Connect with the Client Credentials flow");
+
+        let scope = generate_scope(self.device_id)?;
+
+        let response = self.oidc.authorize_scope_with_client_credentials(scope).await?;
+
+        self.oidc.load_session_meta().await?;
+
+        Ok(response)
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct OidcData {
+    pub(crate) credentials: ClientCredentials,
+    pub(crate) metadata: VerifiedClientMetadata,
+    pub(crate) latest_id_token: RwLock<Option<IdToken<'static>>>,
+}
+
+/// The data needed to perform authorization using OpenID Connect.
+#[derive(Debug, Clone)]
+pub struct OidcAuthorizationData {
+    /// The URL that should be presented.
+    pub url: Url,
+    /// A unique identifier for the request.
+    pub state: String,
+}
+
+fn generate_scope(device_id: Option<String>) -> Result<Scope, OidcError> {
+    let device_id = if let Some(device_id) = device_id {
+        device_id
+    } else {
+        use rand::Rng;
+
+        rand::thread_rng()
+            .sample_iter(&rand::distributions::Alphanumeric)
+            .map(char::from)
+            .take(10)
+            .collect::<String>()
+    };
+
+    Ok([
+        ScopeToken::Openid,
+        ScopeToken::MatrixApi,
+        ScopeToken::try_with_matrix_device(device_id).or(Err(OidcError::InvalidDeviceId))?,
+    ]
+    .into_iter()
+    .collect())
+}
+
+/// The data returned by the provider in the redirect URI after a successful
+/// authorization.
+#[derive(Debug, Deserialize)]
+pub struct AuthorizationResponse {
```

Comment:
I feel like there should be an enum that is either a code or an error, so that the client doesn't have to do the "parse a response or an error" logic itself

I think the logic should be:

```rs
let response = parse_response_from_url(callback_url)?;
client.oidc().process_response(response)?; // This would return an error if the callback returned an error
```

Reply:
`AuthorizationResponse` is now an enum with `Success` and `Error` variants.